### PR TITLE
[WIP] KCRAO client update

### PR DESCRIPTION
**Pull request by tomncooper**
### Type of change

Enhancement / new feature

### Description

Updated KCRAO code to use correct user task endpoint return. Also added basic unit test template for KCRAO.

### Checklist

- [x] Write tests
- [x] Make sure all tests pass

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java]: just a nit, can we use `ccHost` as the member field? :-)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:285]: just because github doesn't help, the only change here is using an expression lambda instead of a statement lambda so removing the "return" right?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:422]: if these cases are about "still processing" we need a `break` before the `default` right? Otherwise, it will go to default and related error.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:552]: ditto as above

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (tomncooper) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:285]: Yes that is correct

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:388]: Being in a `PendingProposal` state means that the operator is waiting for the CC to provide a rebalance proposal after starting a `rebalance` operation with dryrun. As far as I can see from the link, it should be something stoppable, or?
Having a proposal ready, waiting for the user to be approved (`approve` annotation), happens in the `ProposalReady` state and in that case there is nothing to stop of course.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:657]: this branch will actually move the state machine in the `PendingProposal` or `Rebalancing` (depending on dryrun or not) where polling the tasks endpoint happens.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:651]: We can decide that our API client will return an empty or `null` JSON in this case but it's a matter of reaching an agreement and we can change here accordingly.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (tomncooper) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:388]: Well, for that endpoint is says:

> The following POST request will let Kafka Cruise Control stop an ongoing rebalance, add_broker, remove_broker, fix_offline_replica, topic_configuration or demote_broker operation

That list does not include the calculation of a proposal but I will have to check the CC source code to be sure.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperatorTest.java
Comment (ppatierno) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperatorTest.java]: we should try to make the `pendingCalls` parameter configurable on test basis because we want to have different tests which provides the result immediately or after X calls. This is just a comment we should think about, nothing to address right now.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:388]: Isn't it the `rebalance` endpoint? This is the one I am using for asking a rebalance proposal setting the dryrun flag as well.

Comment (tomncooper) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:388]: No, in this case if the stop annotation is present you are calling the ` apiClient.stopExecution` which is a different endpoint to the `rebalance`. As far as I can see there is no way to stop the calculation of a "proposal" once it has started.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:423]: I was stupid I think, I didn't see this `break` here? it's enough for all the above case instead of having a `break` for each of them. The fall through will go to the last break.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:552]: Ditto as above about a single `break`

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:388]: What I meat was that the doc says:

> the following POST request will let Kafka Cruise Control stop an ongoing rebalance...

so it sounds calling `apiClient.stopExecution` should be able to stop an operation started on the `rebalance` endpoint which is the one I use for rebalance proposal.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (tomncooper) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:423]: You aren't stupid, I just pushed a commit to add them ;-) (to fix my stupidity). I confess I am not that familiar with the behavior of Java `switch case` statements. I like being explicit where possible.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (tomncooper) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:388]: I think the we need to make clear the difference between a `rebalance(dryrun=true,customGoals=[goal1, goal2, etc])` which will start the calculation of a "proposal" and doing the above with `dryrun=true` which will start the actual data movements which the `stopExecution` endpoint will stop.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperatorTest.java
Comment (ppatierno) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperatorTest.java:92]: a new `KafkaClusterRebalance` resource doesn't need to have an initial condition with status `New`. Of course, it's something we cannot ask to the user. For this reason when the reconcile starts, it's the operator to set this status to `New` when the condition is null (https://github.com/kyguy/strimzi-kafka-operator/blob/kafka-cluster-rebalance-operator/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java#L605)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:388]: What I understand from the doc is that the `rebalance` endpoint is about triggering a workload balance. When dryrun=true, it provides the proposal for the list of goals you specify. 
Then, as far as I can see it says:

> To let Kafka Cruise Control actually move data, users need to explicitly set dryrun=false

So it means that calling the `rebalance` endpoint with dryrun=false does the actual data movement.

Then, if the doc is right, it seems that the stop execution acts on the `rebalance` endpoint and it's not related to the fact you call it with dryrun=false|true, so it should be able to stop an ongoing proposal processing or actual movement.
Of course, I could have misunderstood the doc :-)

Comment (tomncooper) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:388]: Yes that is all true but this is in the `pendingProposal` method so there is currently no rebalance to stop. So when you are calling `stopExecution` in this method AFAIU it will do nothing?

Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:388]: The `onPendingProposal` method is called when in `PendingProposal` state which means a `rebalance` endpoint was called with dryrun=true and we are waiting for the proposal to be ready so I guess it could be stopped by `stopExecution`.
To be clear, when a `KafkaClusterRebalance` resource is created we are in `New` state and the `onNew` method is called which calls `requestRebalance` which calls the `rebalance` endpoint with dryrun=true. At this point we can get a proposal already ready so we move to `ProposalReady` or, if the proposal is not ready, we move to `PendingProposal`. At this point, it's in the `onPendingProposal` that we check the status or eventually call `stopExecution` so while a rebalance proposal processing is ongoing

Comment (tomncooper) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:388]: Agreed, but calling `stopExecution` before `reabalance(dryrun=false)` won't have any effect as far as I understand it. But I am going over the CC code to make sure.

Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:388]: > stopExecution before reabalance(dryrun=false)

But how can this happen? the `stopExectuion` is called only in the `PendingProposal` state which means you have already called `rebalance` with dryrun=false.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperatorTest.java
Comment (ppatierno) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperatorTest.java]: @tomncooper because this is all about mocking, does it work if we make the various `setupXXXX` methods to be public in the `MockCruiseControl` class?
I mean, if there is a specific test where I need a `pendingCalls` greater than zero, we could just call the `setupCCUserTasksResponse(ccServer, pendingCalls);` in the specific test to change the behavior. Wdyt?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperatorTest.java
Comment (ppatierno) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperatorTest.java:114]: the problem I see mocking in this way, is that we are returning always the initial `kcRebalance` instance while during the reconcile, in the real world, its status is updated with proposal information, session-id, etc etc.
I think it's our first time doing a mock test where the mocked class has to return an updated instance "at some point".

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperatorTest.java
Comment (tomncooper) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperatorTest.java]: Good idea

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperatorTest.java
Comment (tomncooper) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperatorTest.java:114]: Yeah this is the part where I wanted your input on how to proceed.

Comment (ppatierno) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperatorTest.java:114]: Not a mockito expert so I need to think/study more about that :-D

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java
Comment (tomncooper) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:388]: Right, but the `PendingProposal` state represents:
```
       /**
         * A proposal has been requested from Cruise Control, but is not ready yet.
         * Transitions to:
         * <dl>
         *     <dt>ProposalReady</dt><dd>Once Cruise Control has a proposal.</dd>
         *     <dt>NotReady</dt><dd>If Cruise Control returned an error</dd>
         *     <dt>Stopped</dt><dd>If the user sets annotation strimzi.io/rebalance=stop.</dd>
         * </dl>
         */
         PendingProposal,
```
The "proposal" is requested using `dryrun=true`. So we are waiting for CC to tell us what it intends to do. From there we can go to `ProposalReady` and get on with getting approval from the user. Or we can call the whole thing off by the user attaching "stop" to the KafkaRebalance CR. If we are stopping a pending "proposal" there is nothing to "stop" on the CC side, as the user has not approved the proposal so no data is moving, so there is no need to call the stopExecution endpoint? 

My main point is that:
```
if (state(freshClusterRebalance) == State.PendingProposal) {
    if (getRebalanceAnnotation(freshClusterRebalance) == RebalanceAnnotation.stop) {
         apiClient.stopExecution(........)
    }
}
```
Is not needed as a `stopExecution` call when there is no rebalance operation running (PendingProposal) will not actually do anything.

In that particular case we just attach the "stopped" status/annotation and move on?

Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java:388]: > If we are stopping a pending "proposal" there is nothing to "stop" on the CC side, as the user has not approved the proposal so no data is moving.

My understanding is that you can stop the running process of CC to find a proposal. It's not actually moving data of course but it's doing some processing to find a proposal for you. Isn't it stoppable?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/CruiseControlApiImpl.java
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/CruiseControlApiImpl.java:29]: genuine question, what's deprecated here that we need to suppress?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/CruiseControlApiImpl.java
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/CruiseControlApiImpl.java:33]: isn't it too much to log always the entire network activity?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/MockCruiseControl.java
Comment (ppatierno) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/MockCruiseControl.java:70]: instead of deleting all of them here, my idea was about leaving them but being able to override, if needed, in specific tests. Removing them from here, now we need to call at least one `setupXXX` in each test.
I don't know if my idea is going to work because we mock the `ccServer` in the constructor and then we "override" the mock during a test (so for example, reapplying `.when` and so on). Does it cause any problem with mocking? wdyt?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/CruiseControlApiImpl.java
Comment (tomncooper) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/CruiseControlApiImpl.java:29]: The way we use the vertx http client is deprecated. It matches what we do currently for other REST APIs but vertx is deprecating the methods we use for a new `WebClient` class which requires more dependencies. Eventually we will have to refactor to use this new class in all API clients but for now we are supressing the warning.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/CruiseControlApiImpl.java
Comment (tomncooper) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/CruiseControlApiImpl.java:33]: Good point, I think I did that for debugging

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/MockCruiseControl.java
Comment (tomncooper) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/cruisecontrol/MockCruiseControl.java:70]: So we could do that with any of the _expectations_ (what MockServer calls `.when()` calls) that aren't parameterised. However, there is no way to "override" already existing expectations so you would need to `ccServer.reset()` and then apply the new ones. Basically as `.reset()` is a blunt tool you would end up needing to reapply any relevant expectations if `.reset` is called anywhere in your test class. At which point we may as well `.reset()` after every test and apply the expectations you need for each test at the start.

**Review by ppatierno — APPROVED** [review]

**Pull request merged**
